### PR TITLE
[Ready for Review/Testmerge]Adds scanner gates

### PR DESCRIPTION
**Pull request by XDTM**
:cl: XDTM
add: Added scanner gates: these buildable machines will scan any mob passing under it, and it will alert when a set condition is met.
add: Possible conditions: Wanted status, Mindshield, Nanites, Diseases, Species, or carrying guns.
add: Can be inverted, so it will alert when the condition is _not_ met, and can be id-locked to avoid tampering.
/:cl:

Currently ready and tested, but there's room for improvement (signaller integration, directional icons, etc.)

### CODE REVIEW

**Comment by 81Denton** [comment]
I have some ideas:
- Allow attaching signaler assemblies, for example for pulsing airlock wires, bombs, circuitry
- AND/OR/XOR etc. logic gates, for example "alert if lizard with gun but no mindshield passes through this"
- Radiation scanning
- Scanning the ID's assigned job
- Does get_disease_severity_value respect the virus' stealth level?

File: code/datums/diseases/_disease.dm [hunk 1]
```diff
@@ -160,3 +160,21 @@
...
+		if(DISEASE_SEVERITY_MINOR)
+			return 3
+		if(DISEASE_SEVERITY_MEDIUM)
+			return 4
+		if(DISEASE_SEVERITY_HARMFUL)
+			return 5
+		if(DISEASE_SEVERITY_DANGEROUS)
+			return 6
+		if(DISEASE_SEVERITY_BIOHAZARD)
+			return 7
```
Comment (ExcessiveUseOfCobblestone) [inline code/datums/diseases/_disease.dm]: isn't this the purpose of defines...

Comment (XDTM) [reply code/datums/diseases/_disease.dm]: yes but those are text defines

**Comment by MrDoomBringer** [comment]
NICE
i was just playing goon the other day and wishing we had these

**Comment by ExcessiveUseOfCobblestone** [comment]
@81Denton for your last question: yes per https://github.com/tgstation/tgstation/pull/39986/files#diff-98e1663438f2152652cd609c8987315fR97

It's the same proc used for medhuds, PENLITEs, and now scanners

File: code/datums/diseases/_disease.dm (hunk 1 above)
Comment (MrDoomBringer) [reply code/datums/diseases/_disease.dm]: just make a dictionary smhfam

File: code/game/machinery/scan_gate.dm [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+			if(ishuman(M))
+				var/mob/living/carbon/human/H = M
+				if(is_species(H, detect_species))
+					beep = TRUE
+				if(detect_species == /datum/species/zombie) //Can detect dormant zombies
+					if(H.getorganslot(ORGAN_SLOT_ZOMBIE))
+						beep = TRUE
+		if(SCANGATE_GUNS))
+			for(var/I in M.get_contents())
+				if(istype(I, /obj/item/gun))
```
Comment (JJRcop) [inline code/game/machinery/scan_gate.dm]: This will go off with wands and staves, blacklist /obj/item/gun/magic?

Comment (XDTM) [reply code/game/machinery/scan_gate.dm]: I thought it would be a lot of extra processing to do so, so i kept it simple

**Comment by MimicFaux** [comment]
Could this also be made to detect romerol/xenomorph infections?

**Comment by XDTM** [comment]
zombie species detects romerol infections

File: code/game/machinery/scan_gate.dm (hunk 2 above)
Comment (JJRcop) [reply code/game/machinery/scan_gate.dm]: You could create a static typecache of /obj/item/gun, then subtract `typesof(/obj/item/gun/magic)` from the typecache.

File: code/game/machinery/scan_gate.dm [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	
+/obj/machinery/scanner_gate/proc/perform_scan(mob/living/M)
+	var/beep = FALSE
+	switch(scangate_mode)
+		if(SCANGATE_NONE)
+			return
+		if(SCANGATE_WANTED)
+			if(iscarbon(M))
+				var/mob/living/carbon/C = M
+				var/perpname = M.get_face_name(M.get_id_name())
```
Comment (JJRcop) [inline code/game/machinery/scan_gate.dm]: Gotta use C

File: code/game/machinery/scan_gate.dm (hunk 2 above)
Comment (XDTM) [reply code/game/machinery/scan_gate.dm]: that could work, it might allow me to make other types of contraband scans

File: code/datums/diseases/_disease.dm (hunk 1 above)
Comment (ninjanomnom) [reply code/datums/diseases/_disease.dm]: @MrDoomBringer is correct here

File: code/game/machinery/scan_gate.dm [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+																								DISEASE_SEVERITY_MINOR,
+																								DISEASE_SEVERITY_MEDIUM,
+																								DISEASE_SEVERITY_HARMFUL,
+																								DISEASE_SEVERITY_DANGEROUS,
+																								DISEASE_SEVERITY_BIOHAZARD)
+			if(new_threshold)
+				disease_threshold = new_threshold
+			. = TRUE
+		//Some species are not scannable, like abductors (too unknown), androids (too artificial) or skeletons (too magic)
+		if("set_target_species")
```
Comment (F-OS) [inline code/game/machinery/scan_gate.dm]: racist

**Comment by XDTM** [comment]
- Added icon, made by @Nabski89 
- Added a new sound as the alarm
- Added the option to specify a specific nanite cloud ID to check
- Fixed a couple HUD bugs
- FIxed a small bug where captain's ID was missing an access present in the cargo techs' ID (and thus they could lock the scanner without the captain being able to unlock it)

File: code/game/machinery/scan_gate.dm [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+	..()
+	if(!(stat & (BROKEN|NOPOWER)) && isliving(AM))
+		perform_scan(AM)
+
+/obj/machinery/scanner_gate/proc/set_scanline(type, duration)
+	cut_overlays()
+	deltimer(scanline_timer)
+	add_overlay(type)
+	if(duration)
+		scanline_timer = addtimer(CALLBACK(src, .proc/set_scanline, "passive"), duration, TIMER_STOPPABLE)
```
Comment (AnturK) [inline code/game/machinery/scan_gate.dm]: Is there something special simple flick or state change won't do here ? Cutting all overlays seems excessive.

File: code/game/machinery/scan_gate.dm [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+																								SCANGATE_WANTED,
+																								SCANGATE_SPECIES)
+			if(new_mode)
+				scangate_mode = new_mode
+			. = TRUE
+		if("toggle_reverse")
+			reverse = !reverse
+			. = TRUE
+		if("set_disease_threshold")
+			var/new_threshold = input("Set disease threshold","Scan Mode") as null|anything in list(DISEASE_SEVERITY_POSITIVE,
```
Comment (AnturK) [inline code/game/machinery/scan_gate.dm]: You don't do any intractability validation after the inputs here.

File: code/modules/jobs/access.dm [hunk 7]
```diff
@@ -127,7 +127,7 @@
 	            ACCESS_EXTERNAL_AIRLOCKS, ACCESS_CHANGE_IDS, ACCESS_AI_UPLOAD,
 	            ACCESS_TELEPORTER, ACCESS_EVA, ACCESS_HEADS, ACCESS_CAPTAIN, ACCESS_ALL_PERSONAL_LOCKERS,
 	            ACCESS_TECH_STORAGE, ACCESS_CHAPEL_OFFICE, ACCESS_ATMOSPHERICS, ACCESS_KITCHEN,
-	            ACCESS_BAR, ACCESS_JANITOR, ACCESS_CREMATORIUM, ACCESS_ROBOTICS, ACCESS_CARGO, ACCESS_CONSTRUCTION,
+	            ACCESS_BAR, ACCESS_JANITOR, ACCESS_CREMATORIUM, ACCESS_ROBOTICS, ACCESS_CARGO, ACCESS_CARGO_BOT, ACCESS_CONSTRUCTION,
```
Comment (AnturK) [inline code/modules/jobs/access.dm]: Was this something missing or some merge issue ?

File: code/modules/mob/living/carbon/human/species_types/jellypeople.dm [hunk 8]
```diff
@@ -1,6 +1,6 @@
 /datum/species/jelly
 	// Entirely alien beings that seem to be made entirely out of gel. They have three eyes and a skeleton visible within them.
-	name = "Xenobiological Jelly Entity"
+	name = "Jellyperson"
```
Comment (AnturK) [inline code/modules/mob/living/carbon/human/species_types/jellypeople.dm]: Seems completely unrelated.

**Review by AnturK — COMMENTED** [review]
Code looks fine, just some questions.

File: code/modules/mob/living/carbon/human/species_types/jellypeople.dm (hunk 8 above)
Comment (Iamgoofball) [reply code/modules/mob/living/carbon/human/species_types/jellypeople.dm]: name on the scanner stuff prolly

Comment (XDTM) [reply code/modules/mob/living/carbon/human/species_types/jellypeople.dm]: Yeah, i use the species.name for the display name, and this didn't matter before since it was invisible ingame

File: code/modules/jobs/access.dm (hunk 7 above)
Comment (XDTM) [reply code/modules/jobs/access.dm]: Noticed the bug while testing

File: code/game/machinery/scan_gate.dm (hunk 5 above)
Comment (XDTM) [reply code/game/machinery/scan_gate.dm]: I guess i could store the scanline overlay in a var, i was following the example of other update_icon procs

File: code/modules/mob/living/carbon/human/species_types/jellypeople.dm (hunk 8 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/mob/living/carbon/human/species_types/jellypeople.dm]: I actually like the first one better lol

Comment (XDTM) [reply code/modules/mob/living/carbon/human/species_types/jellypeople.dm]: It is nice but base jellypeople are no longer a thing, so it's better to have a catch-all name for jelly species in this case.

File: code/game/machinery/scan_gate.dm (hunk 5 above)
Comment (AnturK) [reply code/game/machinery/scan_gate.dm]: It's not horrible, just could be optimized probably.

**Comment by AnturK** [comment]
Rebuild tgui and this should be good.

**Comment by AnturK** [comment]
@XDTM Still alive ?

**Comment by XDTM** [comment]
sorry, i keep forgetting about reviews
